### ISSUE TITLE
Add benchmarking and regression testing

### ISSUE THREAD

**Issue by Randl** [tiny-dnn/tiny-dnn#548]
Since performance is pretty important for us, we got to make some automated benchmarking and performance testing.
Running on real life example can include Caffe converter testing

**Comment by beru** [comment]
For this purpose, I'd like to modify `examples/cifar10/train.cpp` and `examples/mnist/train.cpp` programs to accept command line arguments for minibatch size and a number of epochs to train. They are fixed to 10 and 30. It's hard for me to wait for more than 5 minutes to measure performance.

And I'm not sure if we can rely on cloud server to measure this library's performance as doing so will consume much computing resource. So how about simply readying a shell script to execute example training programs? Anyone who wants to test performance can run the script locally.

I don't know how to share benchmark records easily. I think we can manually update Wiki for it...

**Comment by Randl** [comment]
@beru I think having regression testing is important, so that we will now that particular PR doesn't reduce performance.

> For this purpose, I'd like to modify examples/cifar10/train.cpp and examples/mnist/train.cpp programs to accept command line arguments for minibatch size and a number of epochs to train.

Agreed

> And I'm not sure if we can rely on cloud services to measure this library's performance as doing so will consume much computing resource. 

I think, running short test to measure CPU time for two epochs (for example) in a couple of examples should be ok.

**Comment by Randl** [comment]
To clarify, I've meant automated benchmarking and comparison between releases, not about comparison between tiny-dnn and other framworks (as in https://github.com/tiny-dnn/tiny-dnn/issues/197 https://github.com/tiny-dnn/tiny-dnn/issues/75 and https://github.com/ctuning/ck-tiny-dnn )

**Comment by beru** [comment]
Running performance check on cloud servers may not be a good idea.
https://github.com/travis-ci/travis-ci/issues/352
I guess we can't guarantee same config hardware will continuously be used.

**Comment by bhack** [comment]
Yes we need to do that on dev machine

**Comment by Randl** [comment]
Maybe add a test with time measurements which prints results and ask devs to post its output together with PR?

**Comment by bhack** [comment]
Yes but we need hardware fingerprint whencs sending cause we need to be sure if we are on the exact same platform.

**Comment by Randl** [comment]
It's enough to show before and after, tho if there are significant changes, additional runs should be performed

**Comment by bhack** [comment]
Execute on master before compile and execute on branch? We need sha.

**Comment by beru** [comment]
> Maybe add a test with time measurements which prints results and ask devs to post its output together with PR?

I fear asking it would raise the bar for creating a pull request. I hope people may voluntarily post benchmark results on comments or Wiki.

**Comment by Randl** [comment]
> I fear asking it would raise the bar for creating a pull request.

Yes it might, but if it will be part of test suite, then it will mean only run tests twice and copy-paste.

**Comment by beru** [comment]
> Yes it might, but if it will be part of test suite, then it will mean only run tests twice and copy-paste.

Running tests twice might be a bit cumbersome, but in that way we can confirm the difference at ease.

Do commiters need to run `cmake` command with `-DBUILD_TESTS=ON` option, and run `tiny_dnn_test` and collect mini-bench result? Sounds tedious but it will make PR a lot cleaner.

If those procedures will be mandatory before sending pull requests, then it should be written in this document.
https://github.com/tiny-dnn/tiny-dnn/blob/master/CONTRIBUTING.md

And members of `tiny-dnn` are also expected to measure performance of pull requests and share the result as a double check.

**Comment by Randl** [comment]
@beru I hope that contributors run tests anyway :)

We can make in not mandatory but strongly recommended, and warn that without test output PR might take more time to merge or not be merged at all.

**Comment by beru** [comment]
@Randl If we strongly recommend commiters to run tests and benchmarks, that attitude should be applied to maintainers who check pull requests. So I hope posting benchmarks will be voluntary act.

Actually, I'm kind of person who skips running tests locally before creating a pull request, and ended up fixing problems later after travis-ci and AppVeyor jobs failed. I admit disciplines are important but isn't it hard?

**Comment by beru** [comment]
Many of the pull requests may not related to performance issues at all. So if that is obvious, I think it's OK to skip running benchmarks. And with some pull requests, it may be difficult to sustain performance with adding features. But we can always fix problems later.

**Comment by Randl** [comment]
@beru I have to run tests locally to make sure it works :) And you yourself run performance tests pretty often.


As for PRs that don't need performance check - surely there are some - typos, code addition, bug fixes, etc. That's what I tried to say by "strongly recommended" - if we see that there is no need in performance test it won't influence PR, but if it's needed we won't merge untill someone else runs performance check.

**Comment by beru** [comment]
@Randl I see, so we can divide the tasks.

**Comment by edgarriba** [comment]
I vote to setup a script to run tests twice and check the difference

**Comment by Randl** [comment]
Script is kinda hard, at least for me. You need to specify which revisions to run. Doesn't seem easier than run tests twice.
